include/oneapi/dpl/internal/distributed_ranges_impl/sp/algorithms/reduce.hpp

```diff
@@ -0,0 +1,190 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_DR_SP_REDUCE_HPP
+#define _ONEDPL_DR_SP_REDUCE_HPP
+
+#include <sycl/sycl.hpp>
+
+#include "oneapi/dpl/async"
+#include "oneapi/dpl/execution"
+#include "oneapi/dpl/numeric"
+
+#include "../../concepts/concepts.hpp"
+#include "../../detail/onedpl_direct_iterator.hpp"
+#include "../init.hpp"
+#include "execution_policy.hpp"
+
+namespace
+{
+
+// Precondition: stdrng::distance(first, last) >= 2
+// Postcondition: return future to [first, last) reduced with fn
+template <typename T, typename ExecutionPolicy, std::bidirectional_iterator Iter, typename Fn>
+auto
+reduce_no_init_async(ExecutionPolicy&& policy, Iter first, Iter last, Fn&& fn)
+{
+    Iter new_last = last;
+    --new_last;
+
+    std::iter_value_t<Iter> init = *new_last;
+
+    oneapi::dpl::experimental::dr::__detail::direct_iterator d_first(first);
+    oneapi::dpl::experimental::dr::__detail::direct_iterator d_last(new_last);
+
+    return oneapi::dpl::experimental::reduce_async(std::forward<ExecutionPolicy>(policy), d_first, d_last,
+                                                   static_cast<T>(init), std::forward<Fn>(fn));
+}
+
+template <typename T, typename ExecutionPolicy, std::bidirectional_iterator Iter, typename Fn>
+requires(sycl::has_known_identity_v<Fn, T>) auto reduce_no_init_async(ExecutionPolicy&& policy, Iter first, Iter last,
```

Comment:
Is this case really needed? I mean is this faster than the version above where we reduce one element less and use it as `init`?

Reply:
Yes, the version without known identity is over 2x slower.  The version that copies the init requires a explicit GPU->CPU copy, which is often as expensive as launching the algorithm itself.

Reply:
I see, thanks.